### PR TITLE
Update documentation for table functions

### PR DESCRIPTION
**Pull request by kasiafi**
## Description

This change adds information about all types of arguments supported by table functions to both developer guide and user guide. Also, the developer guide is updated with the new way of executing table functions involving the operator.

Follow-up to: https://github.com/trinodb/trino/pull/15575

No release notes required. Only docs change.

### CODE REVIEW

File: docs/src/main/sphinx/develop/table-functions.rst [hunk 1]
```diff
@@ -58,11 +58,49 @@ identified by the catalog name, the schema name, and the function name.
...
+
+    new DescribedTable(descriptor)
+
+If a table function does not produce any columns, and it only outputs the
+pass-through columns, use ``ONLY_PASS_THROUGH`` as the returned row type.
+
+Note: a table function must return at least one column. It can either be a
+proper column, i.e. produced by the function, or a pass-through column.
+
+.. _argument_types:
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/table-functions.rst:94]: all ref anchors share a global namespace in the docs .. should we add some prefix for 
this page like _tf_argument_type

File: docs/src/main/sphinx/develop/table-functions.rst [hunk 2]
```diff
@@ -148,12 +238,36 @@ execute the table function invocation:
...
+arguments. To use this execution path, you provide an implementation of a
+processor.
+
+If your table function has one or more table arguments, you should implement
+``TableFunctionDataProcessor``. It processes pages of input data in a work
+processor-based model.
+
+If your table function is a source operator (it does not have table arguments),
+you should implement ``TableFunctionSplitProcessor``. It processes splits in a
+work processor-based model. The connector that provides the function should
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/table-functions.rst:268]:
```suggestion
work processor-based model. The connector that provides the function must
```

File: docs/src/main/sphinx/functions/table.rst [hunk 3]
```diff
@@ -47,6 +47,58 @@ by the specified catalog, the query fails.
...
+    factor => 42
+
+2. Descriptor arguments
+
+Descriptors consist of fields with names and optional data types::
+
+    schema => DESCRIPTOR(id BIGINT, name VARCHAR)
+    columns => DESCRIPTOR(date, status, comment)
+
+To pass ``null`` for descriptor, use::
```
Comment (mosabua) [inline docs/src/main/sphinx/functions/table.rst:69]:
```suggestion
To pass ``null`` for a descriptor, use::
```

File: docs/src/main/sphinx/functions/table.rst [hunk 4]
```diff
@@ -47,6 +47,58 @@ by the specified catalog, the query fails.
...
+
+    schema => CAST(null AS DESCRIPTOR)
+
+3. Table arguments
+
+You can pass a table name, or a query. Use the keyword ``TABLE``::
+
+    input => TABLE(orders)
+    data => TABLE(SELECT * FROM region, nation WHERE region.regionkey = nation.regionkey)
```
Comment (mosabua) [inline docs/src/main/sphinx/functions/table.rst:78]: what the context of that query in terms of catalog, schema..

File: docs/src/main/sphinx/functions/table.rst [hunk 5]
```diff
@@ -47,6 +47,58 @@ by the specified catalog, the query fails.
...
+
+If the table argument is declared as :ref:`set semantics<set_or_row_semantics>`,
+you can specify partitioning and ordering. Each partition is processed
+independently by the table function. If you do not specify partitioning, the
+argument is processed as a single partition.
+
+Additionally, if the table argument is declared as
```
Comment (mosabua) [inline docs/src/main/sphinx/functions/table.rst:85]: do we really want to link again and have a new paragraph.. its about the same "set semantics" in both paragraphs

File: docs/src/main/sphinx/functions/table.rst [hunk 6]
```diff
@@ -47,6 +47,58 @@ by the specified catalog, the query fails.
...
+:ref:`set semantics<set_or_row_semantics>`, you can specify ``PRUNE WHEN EMPTY``
+or ``KEEP WHEN EMPTY``. With ``PRUNE WHEN EMPTY`` you declare that you are not
+interested in the function result if the argument is empty. This information
+is used by the Trino engine to optimize the query. Analogically, the
```
Comment (mosabua) [inline docs/src/main/sphinx/functions/table.rst:89]:
```suggestion
is used by the Trino engine to optimize the query. The
```

File: docs/src/main/sphinx/functions/table.rst [hunk 7]
```diff
@@ -47,6 +47,58 @@ by the specified catalog, the query fails.
...
+``KEEP WHEN EMPTY`` option indicates that the function should be executed even
+if the table argument is empty. Note that by specifying ``KEEP`` or
```
Comment (mosabua) [inline docs/src/main/sphinx/functions/table.rst:91]: Is it really just KEEP here?

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/develop/table-functions.rst [hunk 8]
```diff
@@ -72,26 +110,77 @@ skipped during invocation:
...
+invocation, the user can specify partitioning and ordering for the argument. If
+no partitioning is specified, the argument is processed as a single partition.
+
+A table argument with row semantics is processed on row-by-row basis.
+Partitioning or ordering is not applicable.
+
+Prune or keep when empty
+========================
+
+The ``prune when empty`` property indicates that if the given table argument is
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply docs/src/main/sphinx/develop/table-functions.rst:168]: Actually, in the context of the developer guide, the properties like `prune when empty`, `keep when empty`, and `pass-through columns` don't have official names that you might use anywhere in the code. You set them with a Builder, and this section refers to the builder methods:
```
    TableArgumentSpecification.builder()
            .name("INPUT")
            .rowSemantics()
            .pruneWhenEmpty()
            .passThroughColumns()
            .build()
```
I understand that formatting them as code is confusing. I'll change to emphasis.

File: docs/src/main/sphinx/develop/table-functions.rst [hunk 9]
```diff
@@ -148,12 +238,36 @@ execute the table function invocation:
...
-representing the table function result. The table function invocation is then
-replaced with a ``TableScanNode``.
+There are two paths of execution available for table functions.
+
+1. Pushdown to the connector
+
+The connector that provides the table function implements the
+``applyTableFunction()`` method. This method is called during the optimization
+phase of query processing. It returns a ``ConnectorTableHandle`` and a list of
+``ColumnHandle`` s representing the table function result. The table function
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply docs/src/main/sphinx/develop/table-functions.rst:248]: Without the space, the formatting breaks.

File: docs/src/main/sphinx/develop/table-functions.rst [hunk 10]
```diff
@@ -148,12 +238,36 @@ execute the table function invocation:
...
+invocation is then replaced with a ``TableScanNode``.
+
+This execution path is convenient for table functions whose results are easy to
+represent as a ``ConnectorTableHandle``, for example query pass-through. It
+only supports scalar and descriptor arguments.
+
+2. Execution by operator
+
+Trino has a dedicated operator for table functions. It can handle table
+functions with any number of table arguments as well as scalar and descriptor
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply docs/src/main/sphinx/develop/table-functions.rst:258]: I don't understand this suggestion. The intention of this sentence is to tell that the operator-based execution is available for table functions with table arguments, but also scalar and descriptor arguments are supported. This is in contrast with the preexisting execution path (by pushdown) which supported only scalar and descriptor arguments.

If that is not clear from what I wrote, could you please help me reword?

File: docs/src/main/sphinx/develop/table-functions.rst [hunk 11]
```diff
@@ -148,12 +238,36 @@ execute the table function invocation:
...
+arguments. To use this execution path, you provide an implementation of a
+processor.
+
+If your table function has one or more table arguments, you should implement
+``TableFunctionDataProcessor``. It processes pages of input data in a work
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply docs/src/main/sphinx/develop/table-functions.rst:263]: removed

File: docs/src/main/sphinx/functions/table.rst (hunk 5 above)
Comment (kasiafi) [reply docs/src/main/sphinx/functions/table.rst:85]: merged

File: docs/src/main/sphinx/functions/table.rst (hunk 4 above)
Comment (kasiafi) [reply docs/src/main/sphinx/functions/table.rst:78]: It's the context of the enclosing query. I don't think it needs a comment.

This is not like a query pass-through. In query pass-through, you pass a textual query. Here, you pass the actual data. The query is executed by the engine, and the result is streamed into the TF operator.

File: docs/src/main/sphinx/functions/table.rst (hunk 7 above)
Comment (kasiafi) [reply docs/src/main/sphinx/functions/table.rst:91]: should be ``KEEP WHEN EMPTY``

**Comment by kasiafi** [comment]
@mosabua thanks for your review! I applied or answered comments. Ready for another round.

**Comment by kasiafi** [comment]
@mosabua could you please take another look?

File: docs/src/main/sphinx/develop/table-functions.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (mosabua) [reply docs/src/main/sphinx/develop/table-functions.rst:260]: haha.. dang sphinx .. oh well

File: docs/src/main/sphinx/functions/table.rst (hunk 4 above)
Comment (mosabua) [reply docs/src/main/sphinx/functions/table.rst:78]: ok ..

**Review by mosabua — APPROVED** [review]
Gave it another read. I think this is good to go.

**Pull request merged**